test/system/freertos/sleep.c

```diff
@@ -18,14 +18,14 @@
 static int sleep(void)
 {
 	int rc;
-	unsigned int usec = 3;
+	unsigned int usec = 10000;
 	unsigned long long tstart, tend, tdelayed;
 
 	tstart = metal_get_timestamp();
-	metal_sleep_usec((usec/portTICK_PERIOD_MS));
+	metal_sleep_usec(usec);
 	tend = metal_get_timestamp();
 	tdelayed = tend - tstart;
-	if (tdelayed > (usec/portTICK_PERIOD_MS))
+	if (tdelayed > (usec + 1000) || tdelayed < (usec - 1000))
```

Comment:
I wonder if this test is valid? I'm not sure I understand the value "1000"... Do you expect an accuracy of 1ms ? 
Look to me that tdelayed is a multiple of portTICK_PERIOD_MS so that you should use it in your test to avoid fails on some platforms.

Something like this:
if (tdelayed > (usec) || tdelayed < (usec + portTICK_PERIOD_MS)

Reply:
The value 1000 was an arbitrary choice I made during my early investigation of the problem. You are correct. I will change the test to 
if (tdelayed > (usec + portTICK_PERIOD_MS * 1000) || tdelayed < (usec - portTICK_PERIOD_MS * 1000))
The reason behind the +/- 1 here is that depending on when in the tick period the sleep occurs, and how much usec is rounded by, the elapsed time could be 1 tick more or less than planned.

Reply:
> The value 1000 was an arbitrary choice I made during my early investigation of the problem. You are correct. I will change the test to
> if (tdelayed > (usec + portTICK_PERIOD_MS * 1000) || tdelayed < (usec - portTICK_PERIOD_MS * 1000))
> The reason behind the +/- 1 here is that depending on when in the tick period the sleep occurs, and how much usec is rounded by, the elapsed time could be 1 tick more or less than planned.

I think having an elapsed time lower than the one expected might be no safe. 
i would suggest to redesign the __metal_sleep_usec function, to ensure that the elapsed time is always higher than the expected one. Something like this:

```
static inline int __metal_sleep_usec(unsigned int usec)
{	{
        const TickType_t xDelay = ((usec/1000) / portTICK_PERIOD_MS);
        unsigned int remainder = usec % (1000 * portTICK_PERIOD_MS);

	vTaskDelay(remainder ? xDelay + 1 : xDelay);

	return 0;
}
```

Reply:
Since this is a wrapper around a function provided by FreeRTOS, I don't want to get too complicated. My concern is that we could accidentally always wait 1 tick too long because we try to compensate for something that is already being compensated for.

Reply:
vTaskDelay  seems not accurate. as explain in this post: https://www.freertos.org/FreeRTOS_Support_Forum_Archive/June_2011/freertos_vTaskDelay_problem_4570916.html

"If you ask for a 1 tick delay at a 1 ms tick, you will get somewhere between 0ms and 1ms, if you ask for a 2 tick delay, you will get a delay between 1ms and 2ms.  The low end of the range happens if you are at the end of a tick, and just after you call the delay function the tick occurs. Delays of course can also b extended if other tasks at your priority or higher become ready during the delay, as the delay just affect when this task is eligible to get CPU time, not when it will (unless it is the highest priority task) "

So in this case we have a chance to not respect the requested time. For instance requesting 19 ms for a 10 ms tick can lead to 0 ms in bad case.
That's why I would be more favorable to the idea of waiting 1 notch too much not to accumulate the inaccuracies of the division and vTaskDelay . 

That said your point is valid: This should only  be a wrapper from ms to tick.
Perhaps just implementing vTaskDelay( xDelay + 1 )  for all cases  would be a good compromise to compensate the truncated division?

Reply:
The FreeRTOS documentation confirms that this is the intended behavior. My changes will ensure that vTaskDelay() will never be called with a value of 0, which means something specific to FreeRTOS. I don't want to arbitrarily add 1 or do any rounding to keep things simple.

It seems that this is an area of libmetal needs a deeper look at the design and documentation. Can we use this FreeRTOS-specific change for the release and open an issue for a deeper look across the other implementations?

Reply:
> The FreeRTOS documentation confirms that this is the intended behavior. My changes will ensure that vTaskDelay() will never be called with a value of 0, which means something specific to FreeRTOS. I don't want to arbitrarily add 1 or do any rounding to keep things simple.

The +1 is not arbitrary here it is used to compensate the error introduced by the conversion from ms to tick, so not specific to FreeRTOS but to conversion in tick. This is something quite common for such conversion to respect at minimum the requested time.

In fact you already do it for the 0 value to avoid that vTaskDelay() returns immediately. 
This probably respond to the use case in rpmsg_virtio (https://github.com/OpenAMP/open-amp/blob/master/lib/rpmsg/rpmsg_virtio.c#L322)
RPMSG_TICKS_PER_INTERVAL is equal to 1ms ( with a tick at 5 or  10 ms).
So what i proposed here was to not add an exception for 0 only but to compensate for all the requested values.

That's said, examples in FreeRTOS do not add 1 to the conversion result.
So if you are still not convince ,  OK to merge this.

> It seems that this is an area of libmetal needs a deeper look at the design and documentation. Can we use this FreeRTOS-specific change for the release and open an issue for a deeper look across the other implementations?

Reply:
I would prefer not to add 1 in all cases, since that could result in the function taking 1 extra clock tick.